op.c

```diff
@@ -10233,6 +10234,62 @@ Perl_newFOROP(pTHX_ I32 flags, OP *sv, OP *expr, OP *block, OP *cont)
         }
         else if (sv->op_type == OP_NULL && sv->op_targ == OP_SREFGEN)
             NOOP;
+        else if (sv->op_type == OP_LIST) {
+            LISTOP *list = (LISTOP *) sv;
+            OP *pushmark = list->op_first;
+            OP *first_padsv;
+            UNOP *padsv;
+            PADOFFSET i;
+
+            iterpflags = OPpLVAL_INTRO; /* for my ($k, $v) () */
+
+            if (!pushmark || pushmark->op_type != OP_PUSHMARK) {
+                /* diag_listed_as: Can't use %s for loop variable */
+                Perl_croak(aTHX_ "Can't use list for loop variable (not a pushmark)");
+            }
+            first_padsv = OpSIBLING(pushmark);
+            if (!first_padsv || first_padsv->op_type != OP_PADSV) {
+                /* diag_listed_as: Can't use %s for loop variable */
+                Perl_croak(aTHX_ "Can't use list for loop variable (not a padsv)");
+            }
+            padoff = first_padsv->op_targ;
+
+            /* There should be at least one more PADSV to find, and the ops
+               should have consecutive values in targ: */
+            padsv = (UNOP *) OpSIBLING(first_padsv);
+            do {
+                if (!padsv || padsv->op_type != OP_PADSV) {
+                    /* diag_listed_as: Can't use %s for loop variable */
+                    Perl_croak(aTHX_ "Can't use list for loop variable (not a pad2v at %lu)", how_many_more);
```

Comment:
I'm not sure how I feel about these errors, they seem kind of scary to perl users who accidentally do:
```
alh@emondsfield:~/p5/perl$ ./perl -Ilib -e 'for my ($a,,$b) (@list) { }'
for my (...) is experimental at -e line 1.
Can't use list for loop variable (not a pad2v at 0) at -e line 1.
```
Do we usually include internal jargon for easy to trigger parsing errors?

Reply:
Right, in all the other comments I missed the detail of this one

1. we *do* generate panic messages in various places, but they start "panic: "
2. we ought to be able to parse this, and I don't know how...

So this needs a testcase and fixing.

Reply:
How - like this:

```
--- a/perly.y
+++ b/perly.y
@@ -1384,7 +1384,7 @@ my_scalar:        scalar

 /* A list of scalars for "for my ($foo, $bar) (@baz)"  */
 list_of_scalars:       list_of_scalars[list] PERLY_COMMA
-                       { parser->in_my = 0; $$ = $list; }
+                       { $$ = $list; }
        |               list_of_scalars[list] PERLY_COMMA scalar
                        {
                          $$ = op_append_elem(OP_LIST, $list, $scalar);
diff --git a/t/op/for-many.t b/t/op/for-many.t
```

ie that was my bug in the grammar.

What had been happening (the panic) was because the internal state had switched back to `parser->in_my` false at the (first) double comma, and so the subsequent variable(s) optree(s) were built as GV lookups, instead of pad lookups. And when those OPs arrived in newFORLOOP, the paranoia checks fired.

So, branch now should be good in 390937f4796971f94123b325d7c484150be6b04c

Reply:
Awesome, thanks